### PR TITLE
Prevent some exceptions

### LINKED ISSUE #1495: workspace/didChangeConfiguration: InvalidOperationException: An attempt was made to transition a task to a final state when it had already completed.
https://github.com/PowerShell/PowerShellEditorServices/blob/a12edc99e24818ccbd2de9200962fb7eb96104fc/src/PowerShellEditorServices/Extensions/EditorObject.cs#L140

Consider using TrySetResult to prevent this exception when the editor is re-initialized or sends a workspace/didChangeConfiguration request

@rjmholt

### PR DESCRIPTION
**Pull request by andschwa**
Fixes #1495, #1498, and #1476.

### CODE REVIEW

File: src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs [hunk 1]
```diff
@@ -223,9 +223,15 @@ public async Task<string> GetCommentHelpText(string functionText, string helpLoc
...
+        public async Task<IReadOnlyDictionary<string, MarkerCorrection>> GetMostRecentCodeActionsForFileAsync(DocumentUri uri)
         {
-            if (!_mostRecentCorrectionsByFile.TryGetValue(scriptFile, out CorrectionTableEntry corrections))
+            ScriptFile file = null;
+            if (!_workspaceService.TryGetFile(uri, out file))
+            {
+                return null;
+            }
+
+            if (!_mostRecentCorrectionsByFile.TryGetValue(file, out CorrectionTableEntry corrections))
```
Comment (rjmholt) [inline src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs:234]:
```suggestion
            if (!_workspaceService.TryGetFile(uri, out ScriptFile file)
                || !_mostRecentCorrectionsByFile.TryGetValue(file, out CorrectionTableEntry corrections))
            {
                return null;
            }
```

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CodeActionHandler.cs [hunk 2]
```diff
@@ -52,13 +52,12 @@ public override async Task<CommandOrCodeActionContainer> Handle(CodeActionParams
...
-                _logger.LogDebug("CodeAction request canceled at range: {0}", request.Range);
+                _logger.LogDebug($"CodeAction request canceled at range: {request.Range}");
                 return Array.Empty<CommandOrCodeAction>();
             }
 
-            // On Windows, VSCode still gives us file URIs like "file:///c%3a/...", so we need to escape them
-            IReadOnlyDictionary<string, MarkerCorrection> corrections = await _analysisService.GetMostRecentCodeActionsForFileAsync(
-                _workspaceService.GetFile(request.TextDocument.Uri)).ConfigureAwait(false);
+            IReadOnlyDictionary<string, MarkerCorrection> corrections = await
+                _analysisService.GetMostRecentCodeActionsForFileAsync(request.TextDocument.Uri).ConfigureAwait(false);
```
Comment (rjmholt) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CodeActionHandler.cs:61]:
```suggestion
            IReadOnlyDictionary<string, MarkerCorrection> corrections = await _analysisService.GetMostRecentCodeActionsForFileAsync(
                request.TextDocument.Uri)
                .ConfigureAwait(false);
```

File: src/PowerShellEditorServices/Services/Workspace/Handlers/ConfigurationHandler.cs [hunk 3]
```diff
@@ -157,6 +157,12 @@ public override async Task<Unit> Handle(DidChangeConfigurationParams request, Ca
         private void SendFeatureChangesTelemetry(LanguageServerSettingsWrapper incomingSettings)
         {
             var configChanges = new Dictionary<string, bool>();
+            if (incomingSettings == null)
+            {
+                this._logger.LogTrace("Incoming settings were null");
+                return;
+            }
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices/Services/Workspace/Handlers/ConfigurationHandler.cs:165]: Can prevent an unnecessary allocation here:

```suggestion
            if (incomingSettings is null)
            {
                this._logger.LogTrace("Incoming settings were null");
                return;
            }

            var configChanges = new Dictionary<string, bool>();
```

**Comment by andschwa** [comment]
@rjmholt Applied.

File: src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs [hunk 4]
```diff
@@ -223,9 +223,13 @@ public async Task<string> GetCommentHelpText(string functionText, string helpLoc
...
         /// <returns>A threadsafe readonly dictionary of the code actions of the particular file.</returns>
-        public async Task<IReadOnlyDictionary<string, MarkerCorrection>> GetMostRecentCodeActionsForFileAsync(ScriptFile scriptFile)
+        public async Task<IReadOnlyDictionary<string, MarkerCorrection>> GetMostRecentCodeActionsForFileAsync(DocumentUri uri)
         {
-            if (!_mostRecentCorrectionsByFile.TryGetValue(scriptFile, out CorrectionTableEntry corrections))
+            if (!_workspaceService.TryGetFile(uri, out ScriptFile file)
+                || !_mostRecentCorrectionsByFile.TryGetValue(file, out CorrectionTableEntry corrections))
+            {
+                return null;
+            }
```
Comment (rjmholt) [inline src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs:232]: Whoops, this might be my fault

```suggestion
```

Comment (rjmholt) [reply src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs:232]: (There's an identical block underneath)

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CodeActionHandler.cs [hunk 5]
```diff
@@ -52,13 +53,13 @@ public override async Task<CommandOrCodeActionContainer> Handle(CodeActionParams
         {
             if (cancellationToken.IsCancellationRequested)
             {
-                _logger.LogDebug("CodeAction request canceled at range: {0}", request.Range);
+                _logger.LogDebug($"CodeAction request canceled at range: {request.Range}");
                 return Array.Empty<CommandOrCodeAction>();
             }
 
-            // On Windows, VSCode still gives us file URIs like "file:///c%3a/...", so we need to escape them
             IReadOnlyDictionary<string, MarkerCorrection> corrections = await _analysisService.GetMostRecentCodeActionsForFileAsync(
```
Comment (dkattan) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CodeActionHandler.cs:60]: Can we change this to:
```
IReadOnlyDictionary<string, MarkerCorrection> corrections = null;
            try
            {
                var file = _workspaceService.GetFile(request.TextDocument.Uri);
                corrections = await _analysisService.GetMostRecentCodeActionsForFileAsync(file).ConfigureAwait(false);
            }
            catch (FileNotFoundException ex)
            {
                _logger.LogWarning("textDocument/codeAction FileNotFound {0}", request.TextDocument.Uri);
                return Array.Empty<CommandOrCodeAction>();
            }
```

Comment (andschwa) [reply src/PowerShellEditorServices/Services/TextDocument/Handlers/CodeActionHandler.cs:60]: That change was made to `GetMostRecentCodeActionsForFileAsync` itself in fcf2ee7d8221a1ffe883ed54ea4542228515dd9b

File: src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs (hunk 4 above)
Comment (andschwa) [reply src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs]: Heh, oh GitHub. Fixed.

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CodeActionHandler.cs (hunk 5 above)
Comment (dkattan) [reply src/PowerShellEditorServices/Services/TextDocument/Handlers/CodeActionHandler.cs:60]: Ah perfect, thanks!

**Review by rjmholt — APPROVED** [review]

**Pull request merged**